report: add treemap button, refactor icon styles

connorjclark opened the pull request:
Only shows if there is treemap data present.

We can also use this new icon button in devtools for the "View Trace" button.

ref #11254

### tool menu before

![image](https://user-images.githubusercontent.com/4071474/116304974-5df57b00-a760-11eb-9d19-39c27b2b352f.png)
![image](https://user-images.githubusercontent.com/4071474/116305025-6d74c400-a760-11eb-80b4-4921f5b20601.png)


## second pass

![image](https://user-images.githubusercontent.com/4071474/116306623-8a11fb80-a762-11eb-87e1-64d4af5e6c57.png)
![image](https://user-images.githubusercontent.com/4071474/116306658-94cc9080-a762-11eb-96ab-1b39a7a5a334.png)



## first pass

![image](https://user-images.githubusercontent.com/4071474/115637465-716e9500-a2cd-11eb-8e5e-d2ac86a44e45.png)

hover

![image](https://user-images.githubusercontent.com/4071474/115637479-77647600-a2cd-11eb-8b84-a5eac9107bed.png)

adamraine commented on lighthouse-core/report/html/renderer/report-ui-features.js:
```diff
@@ -151,6 +151,16 @@ class ReportUIFeatures {
+    if (showTreemapApp) {
+      this.createIconButton({
+        name: 'treemap',
+        text: 'Open Treemap App',
```
Continuing with trying to match "View Trace", would "View Treemap" also work?

adamraine reviewed (commented):
> We can also use this new icon button in devtools for the "View Trace" button.

Could we get the shape and style of this new button to match "View Trace"? Or are you planning to change "View Trace" down the line when treemap moves into the default config?

connorjclark commented:
> Could we get the shape and style of this new button to match "View Trace"? Or are you planning to change "View Trace" down the line when treemap moves into the default config?

![image](https://user-images.githubusercontent.com/4071474/115771816-9f0a1b80-a36b-11eb-8da6-ab6cce7ebba0.png)
yeah i'll tweak the icon button a bit.

I planned to change the view trace button to use this icon button next release.

paulirish commented on lighthouse-core/report/html/report-styles.css:
```diff
@@ -792,7 +792,32 @@
+  height: 100%;
+  background-size: cover;
+}
+.icon-button--treemap .icon-button__preview {
```
instead of a unique dom element for the icon you could set it as bg image for the entire thing.. same approach we use in the `lh-tools__dropdown`

adding on, we do all sorts of things in there, like `text-indent` to make room for the icon, etc.. i could see us reusing a lot of the same styling technique.

paulirish commented on lighthouse-core/report/html/renderer/report-ui-features.js:
```diff
@@ -169,6 +179,20 @@ class ReportUIFeatures {
+  createIconButton(opts) {
+    const metricsEl = this._dom.find('.lh-audit-group--metrics', this._document);
+    const buttonEl =
+      this._dom.createChildOf(metricsEl, 'div', `icon-button icon-button--${opts.name}`);
```
we gotta [Just use a button](https://www.youtube.com/watch?v=CZGqnp06DnI)

tldr:  `<button type=button`>

paulirish commented on lighthouse-core/report/html/report-styles.css:
```diff
@@ -792,7 +792,32 @@
+  box-shadow: 1px 1px 3px #ccc;
+}
+.icon-button:hover {
+  background-color: var(--color-gray-200);
```
this is correct for light mode, but in dark mode (with the filter invert), this is no bueno.

`#BDBDBD` aka `--color-gray-400` is what we use in the dropdown that inverts to the right shade.

connorjclark replied:
this filter invert made using colors for other parts of the element hot trash, so I did a lil refactor of the icons

paulirish reviewed (commented):
> > Could we get the shape and style of this new button to match "View Trace"? Or are you planning to change "View Trace" down the line when treemap moves into the default config?
> 
> ![image](https://user-images.githubusercontent.com/4071474/115771816-9f0a1b80-a36b-11eb-8da6-ab6cce7ebba0.png)
> we could tweak the icon button a bit to match this. wdyt @paulirish

yup lg!  i was actually also tweaking the look and my first pass ended up with something quite similar:

![image](https://user-images.githubusercontent.com/39191/115791550-7723b380-a37d-11eb-9e75-2d5f36ea9f16.png)

connorjclark commented:
Updated first post with new screenshots. I refactored `tool-icon` (renamed `report-icon`) to be reusable, and for the invert trick we use for the icon to not bleed over into other parts of the element (ty psuedoelements)

paulirish commented on lighthouse-core/report/html/renderer/report-ui-features.js:
```diff
@@ -169,6 +179,25 @@ class ReportUIFeatures {
+  /**
+   * @param {{text: string, icon?: string, onClick: () => void}} opts
+   */
+  createButton(opts) {
```
`addButton` since it adds it to the dom? (from the name i didnt expect it to do the append too, but it makes sense)

paulirish commented on lighthouse-core/report/html/report-styles.css:
```diff
@@ -404,21 +408,47 @@
+  background-image: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 0 24 24" width="24px" fill="black"><path d="M0 0h24v24H0V0z" fill="none"/><path d="M3 5v14h19V5H3zm2 2h15v4H5V7zm0 10v-4h4v4H5zm6 0v-4h9v4h-9z"/></svg>');
+}
+
+.lh-button {
```
the component approach of the reusable report-icon in both dropdown items and lh-button makes sense after you think about it for a bit.

so.

yeah sure. cool. 👍

paulirish commented on lighthouse-core/report/html/templates.html:
```diff
@@ -292,7 +292,7 @@
     .lh-tools__button.active + .lh-tools__dropdown {
       opacity: 1;
-      clip: rect(-1px, 187px, 242px, -3px);
+      clip: rect(-1px, 194px, 242px, -3px);
```
this clip is so weird. whyd we do it? ooh for a composited animation. geez.

paulirish commented on lighthouse-core/report/html/templates.html:
```diff
@@ -324,22 +323,6 @@
-    .dark .lh-tools__dropdown a:focus {
-      background-color: #BDBDBD;
-    }
     /* copy icon needs slight adjustments to look great */
```
this selector doesnt apply anymore fwiw.

it seemed weird so i checked it out. it seems to be shrinking the icon to not be so big.

![image](https://user-images.githubusercontent.com/39191/116475900-4e913300-a82f-11eb-9b5e-8d11a122a126.png)

IMO its not a huge deal so we can just nuke the rule. (or move it to `.report-icon--copy::before`)

paulirish reviewed (approved):
a few small things first, but seems good!
